File: 백준/Silver/16493. 최대 페이지 수/최대 페이지 수.js
const input = require("fs").readFileSync("/dev/stdin").toString().trim().split("\n")

const [N,M] = input[0].split(' ').map(Number)
const books = input.slice(1).map(row=>row.split(' ').map(Number))

const dp = Array.from({length:N+1},()=>0)

for(let i=0;i<M;i++){

  const [day,page] = books[i]

  for(let j=N;j>0;j--){
    if(j>=day){
      dp[j] = Math.max(dp[j],dp[j-day]+page)
    }
  }
}

console.log(Math.max(...dp))

File: 백준/Silver/16493. 최대 페이지 수/최대 페이지 수.py
import sys
input = sys.stdin.readline

N,M = map(int,input().split())
books = [list(map(int,input().split())) for _ in range(M)]

dp = [0] *(N+1)

for i in range(M):
  day,page = books[i]

  for j in range(N,0,-1):
    if(j>=day):
      dp[j] = max(dp[j],dp[j-day]+page)
print(max(dp))